### PR TITLE
[Security] Fix UserNotFoundException is not thrown

### LINKED ISSUE #45070: Security - Custom login authenticator
### Symfony version(s) affected

5.4

### Description

In documentation here https://symfony.com/doc/5.4/security/custom_authenticator.html is written:

> You can optionally pass a user loader as second argument to the UserBadge. This callable receives the $userIdentifier and must return a UserInterface object (otherwise a UserNotFoundException is thrown):

```
return new Passport(
    new UserBadge($email, function ($userIdentifier) {
        return $this->userRepository->findOneBy(['email' => $userIdentifier]);
    }),
    $credentials
);
```
Problem is that UserNotFoundException is not thrown, but AuthenticationServiceException is instead, which results in weird message when logging in "Authentication request could not be processed due to a system problem." instead of simple "Invalid credentials".

For example this works as expected:
```
return new Passport(
    new UserBadge($email),
    $credentials
);
```

### How to reproduce

Create custom login authenticator and compare different behaviour between new UserBadge when no $userLoader is provided and when $userLoader is provided while trying to login with non-existing user email address.

### Possible Solution

_No response_

### Additional Context

```
Symfony\Component\Security\Core\Exception\AuthenticationServiceException: The user provider must return a UserInterface object, "null" given. in ...\vendor\symfony\security-http\Authenticator\Passport\Badge\UserBadge.php:69
```

### PR DESCRIPTION
**Pull request by damienfa**
| Q             | A
| ------------- | ---
| Branch?       |  5.4
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Fix #45070
| License       | MIT
| Doc PR        |  N/A

### CODE REVIEW

File: src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php [hunk 1]
```diff
@@ -66,6 +67,13 @@ public function getUser(): UserInterface
             }
 
             $user = ($this->userLoader)($this->userIdentifier);
+            
+            // No user has been found via the $this->userLoader callback.
+            if (is_null($user)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php]: `if (null === $user) {`

**Review by nicolas-grekas — COMMENTED** [review]
Can you please rebase to target 5.4, and add a test case?

**Comment by nicolas-grekas** [comment]
Can you please add a test case, and rebase + target 5.4 also?

**Comment by kaznovac** [comment]
don't have time to thoroughly check the pr and how does the `UserNotFoundException` propagate, pay attention not to reintroduce the [user enumeration](https://github.com/symfony/symfony/security/advisories/GHSA-3r3h-g39f-v648)
also take in consideration the `hideUserNotFoundExceptions` parameter

**Comment by fabpot** [comment]
@wouterj Can you have a look at this PR?

File: src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php [hunk 2]
```diff
@@ -66,6 +67,13 @@ public function getUser(): UserInterface
             }
 
             $user = ($this->userLoader)($this->userIdentifier);
+            
+            // No user has been found via the $this->userLoader callback.
+            if (is_null($user)) { 
+                ($exception = new UserNotFoundException())->setUserIdentifier($this->userIdentifier);
+                throw $exception;
```
Comment (wouterj) [inline src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php]:
```suggestion
            if (null === $user) { 
                $exception = new UserNotFoundException();
                $exception->setUserIdentifier($this->userIdentifier);

                throw $exception;
```

**Review by wouterj — APPROVED** [review]
With the suggested CS changes, this change looks good to me (for 5.4+).

This code is within the user enumeration control of the authenticator manager - so nothing to worry about concerning that.

**Review by chalasr — APPROVED** [review]
Once CS issues have been adressed

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @damienfa.

**Comment by Nessnkr** [comment]
Hello @fabpot ,
Shouldn't it be merged into 5.4?

**Comment by nicolas-grekas** [comment]
Cherry-picked on 5.4 as 3ed91c4cce
Thanks for the notice @Nessnkr 
It'd be great to have a test case by the way.
Anyone up to send one?

**Comment by chalasr** [comment]
See #46063 for the missing test case

**Pull request merged**